### ISSUE TITLE
Full Interaction History

### ISSUE THREAD

**Issue by meatballs** [Axelrod-Python/Axelrod#284]
At the moment, we score the interactions between players within the RoundRobin class and return the payoff and cooperation matrix. The detail of the interactions themselves is lost.

Should we have an option to included the full history as an additional entry in the dictionary produced by RoundRobin?

**Comment by marcharper** [comment]
How much memory would that take up? I suppose it could be stored efficiently with a bitstring.

**Comment by meatballs** [comment]
Not sure - but I'd want to use a flag, set with a default of False, to control whether it's returned or not

**Comment by uglyfruitcake** [comment]
That would be really useful for analysing and improving strategies. Definitely like the sound of that.

**Comment by drvinceknight** [comment]
I very much like this idea (obviously setting the flag by default to False).

**Comment by marcharper** [comment]
If you all haven't looked at my [AxelrodExamples](https://github.com/marcharper/AxelrodExamples) repository lately, check it out. It basically reproduces the intent of this feature already, producing many plots that are useful for analyzing strategies, such as the following which shows how often Backstabber cooperations in each round with each opponent:

![](https://github.com/marcharper/AxelrodExamples/blob/master/assets/cooperation_heatmaps/BackStabber.png)

This one shows how often opponents cooperate with BackStabber:

![](https://github.com/marcharper/AxelrodExamples/blob/master/assets/opponent_cooperation_heatmaps/BackStabber.png)

There are also similar heatmaps for the mean score per round, mean score difference, and for noisy tournaments. I've been using these to develop some new strategies already (which I haven't submitted yet). I'm happy to write up some articles on all this (it's on my todo list, which has grown quickly recently).

Regardless, I'm not opposed to reproducing (or moving) any of this functionality inside Axelrod, but note that to generate these plots I did not run many tournaments over and over, rather I ran many direct play simulations. My vote would be to work on #248 first -- it's more closely attached to the tournament outcomes and would also be very useful for developing new strategies. It's also one of the few things that we need to get serious about writing some papers (the other big one being the rest of the strategies from the two Axelrod tournaments).

**Comment by meatballs** [comment]
I agree that #248 and the remaining original strategies are higher priorities.

**Comment by drvinceknight** [comment]
:+1:

**Comment by meatballs** [comment]
I also think there's a strong case that the RoundRobin class should only return the interaction history. Computing the payoff and cooperation matrices fit more naturally with the ResultSet class - along with all the other computation methods.

**Comment by marcharper** [comment]
I agree with @meatballs ^

**Comment by langner** [comment]
With 100 strategies there will be around a million rounds in a tournament, so there should not be any issues with memory, using bytes or otherwise:
```python
>>> sys.getsizeof(100*100*[random.choice(['C','D']) for i in range(200)])
16000072
```

**Comment by drvinceknight** [comment]
> I agree with @meatballs ^

:+1: 

Me too :)

**Comment by meatballs** [comment]
If RoundRobin is only going to return the interaction history, then the calculation of the payoff and cooperation matrices need to move elsewhere. A couple of options spring to mind:

1) Create new Payoff and Cooperation classes. Each has an ```__init__``` method which will take the interaction history. Put all the methods from ResultSet which use the payoff or cooperation matrices into those new classes.

2) Move the calculation of the payoff and cooperation matrices into ResultSet and change its ```__init__``` to use the interaction history.

Option (1) feels nicer to me. Anyone got any other thoughts?

**Comment by marcharper** [comment]
In option (1), will all the calculations still end up in ResultSet, including the Payoff and Cooperation matrices?

**Comment by meatballs** [comment]
That would be my plan - ResultSet would initiate the calculations within the new classes and hold the results in its existing attributes. The methods would move but the attributes stay the same.

**Comment by drvinceknight** [comment]
Just a minor comment/thought before I go back to watching the rugby. I
haven't read properly but seen a suggestion of a new class (here and in the
co-op count issue). I think we should perhaps make sure we think carefully
about adding new classes as it does add to the complexity of the code base.
Let's make sure we do it when it's really needed (which might well be the
case here).

**Comment by meatballs** [comment]
In this case, it wouldn't add very much at all to the total code base. It would simply move many of the methods out of the ResultSet class and into two new classes - probably as class methods which would speed things up a little.

The main reason is that I think the ResultSet class is becoming a little too large and unwieldy already. If I add in the payoff and cooperation matrices too, it becomes even worse.

**Comment by meatballs** [comment]
Also, in my opinion, extra classes do not add to the complexity of the code base if they make each individual class more understandable.

**Comment by marcharper** [comment]
I agree with @meatballs -- more classes can be better if they are well-designed and encapsulate behaviors appropriately. Of course this isn't Java, not everything needs to be a class.

The example in #334 is a perfect example. The code is "complex" or "messy" because the `Player` class is having to micromanage the history and that task is spread out over multiple methods. As its own class, the management can be more cleanly encapsulated, and only the necessary bits exposed to other parts of the code.

**Comment by meatballs** [comment]
> Of course this isn't Java, not everything needs to be a class.

Yep - another option would be simply to have a set of functions to handle the payoff and morality calculations. Even if I put them in a class, they would be class methods as there's no state to maintain.

**Comment by marcharper** [comment]
In any case, I agree these calculations are not RoundRobin's responsibility; whether they fit in ResultsSet or new classes I'll leave up you @meatballs.

**Comment by drvinceknight** [comment]
I'm also happy with whatever you prefer :)

**Comment by drvinceknight** [comment]
I have no idea what voodoo did that: https://www.dropbox.com/s/bw0hgdoldqrxqgm/Screenshot%202015-11-28%2008.01.32.png?dl=0

**Comment by meatballs** [comment]
The next step on this road is, I reckon, to replace most of the logic within the RoundRobin class with the creation of Match objects and calls to their play methods, passing the resulting interactions to ``payoff.payoff_matrix``` and ```cooperation.cooperation_matrix```

Do we still want that in a RoundRobin class or should we incorporate that logic into Tournament class?

**Comment by drvinceknight** [comment]
> he next step on this road is, I reckon, to replace most of the logic within the RoundRobin class with the creation of Match objects and calls to their play methods, passing the resulting interactions to payoff.payoff_matrixandcooperation.cooperation_matrix`

Awesome. This is inline with what I've done with the `Match` class for probabilistic games :)

> 
> Do we still want that in a RoundRobin class or should we incorporate that logic into Tournament class and get rid of RoundRobin altogether?

I think I'd vote for keeping the `RoundRobin` is worth doing. I can imagine a couple of instance where perhaps someone might want to modify what a round robin is. Even if it's not going to be a contribution to the actual library someone could perhaps want to create a tournament where a round robin does not involve everyone interacting with everyone else.

**Comment by meatballs** [comment]
> someone could perhaps want to create a tournament where a round robin does not involve everyone interacting with everyone else.

Surely that's not a round robin? That's a completely different type of tournament. If we want to go down that road, we really should get rid of RoundRobin and use a 'type' attribute in the Tournament class to control what sets of Match objects are created.

**Comment by drvinceknight** [comment]
> Surely that's not a round robin? That's a completely different type of tournament. If we want to go down that road, we really should get rid of RoundRobin and use a 'type' attribute in the Tournament class to control what sets of Match objects are created.

Yeah I think you're right actually. Completely agree with you. Happy to get rid of Round Robin. :+1:

**Comment by meatballs** [comment]
How about this:

We keep the RoundRobin class for now and I replace the logic within it. Once we've done that and can see how it looks, we can decide whether to take it any further.

**Comment by drvinceknight** [comment]
Closed by #474
